PubNub/Network/Parsers/PNSubscribeParser.m

```diff
@@ -152,72 +166,54 @@ + (NSDictionary *)parsedServiceResponse:(id)response withData:(NSDictionary *)ad
-                NSString *objectName = ([groups count] > eventIdx ? groups[eventIdx] : nil);
-                NSMutableDictionary *event = [self eventFromData:feedEvents[eventIdx]
-                                                      forChannel:(objectName?: objectOrGroupName)
-                                                           group:(objectName? objectOrGroupName: nil)
-                                        withAdditionalParserData:additionalData];
+                NSMutableDictionary *event = [self eventFromData:feedEvents[eventIdx] withAdditionalParserData:additionalData];
                 event[@"timetoken"] = timeToken;
+                event[@"region"] = region;
                 [events addObject:event];
             }
             feedEvents = [events copy];
         }
-        processedResponse = @{@"events":feedEvents,@"timetoken":timeToken};
+        processedResponse = @{@"events":feedEvents,@"timetoken":timeToken,@"region":region};
     }
-    
+
     return processedResponse;
 }
 
 
 #pragma mark - Events processing
 
-+ (NSMutableDictionary *)eventFromData:(id)data forChannel:(NSString *)channel
-                                 group:(NSString *)group
-              withAdditionalParserData:(NSDictionary *)additionalData {
++ (NSMutableDictionary *)eventFromData:(id)data withAdditionalParserData:(NSDictionary *)additionalData {
     
     NSMutableDictionary *event = [NSMutableDictionary new];
-    if ([channel length]) {
-        
-        event[(![group length] ? @"subscribedChannel": @"actualChannel")] = channel;
+    NSString *channel = data[kPNMessageChannelKey];
+    NSString *subscriptionMatch = data[kPNMessageSubscriptionMatchKey];
+    event[@"subscribedChannel"] = subscriptionMatch;
+    event[@"actualChannel"] = channel;
+    if (data[kPNEventsPublishTimeTokenElementKey] && [data[kPNEventsPublishTimeTokenElementKey] isKindOfClass:[NSDictionary class]]) {
+        event[@"publishTimetoken"] = data[kPNEventsPublishTimeTokenElementKey];
     }
-    if ([group length]) {
-        
-        event[@"subscribedChannel"] = group;
+    if (data[kPNEventsUserMetadataElementKey] && [data[kPNEventsUserMetadataElementKey] isKindOfClass:[NSDictionary class]]) {
+        event[@"userMetadata"] = data[kPNEventsUserMetadataElementKey];
     }
-    
     BOOL isPresenceEvent = [PNChannel isPresenceObject:channel];
-    if (![channel length] && [data isKindOfClass:[NSDictionary class]]) {
-        
-        isPresenceEvent = (data[@"timestamp"] != nil &&
-                           (data[@"action"] != nil || data[@"occupancy"] != nil));
+    if ([data[@"d"] isKindOfClass:[NSDictionary class]]) {
+        NSDictionary *payload = data[@"d"];
+        isPresenceEvent = (payload[@"timestamp"] != nil &&
+                           (payload[@"action"] != nil || payload[@"occupancy"] != nil));
     }
-    
     if (isPresenceEvent) {
-        
         [event addEntriesFromDictionary:[self presenceFromData:data]];
```

Comment:
@jzucker2 Please pass _data[kPNMessagePayloadDataKey]_ instead of hole event envelop (_data_). This method don't care about outer data structure and should work only with presence data information.